src/tls/cockpit-certificate-ensure.c

```diff
@@ -0,0 +1,64 @@
+#include <assert.h>
+#include <err.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <common/cockpitwebcertificate.h>
+
+#include "certificate.h"
+#include "utils.h"
+
+// Renew certificates with less than 30 days validity
+#define EXPIRY_THRESHOLD (30 * 24 * 60 * 60)
+
+static bool
+check_expiry (const char *filename)
+{
+  Certificate *certificate = certificate_load (filename);
+
+  time_t expires = certificate_get_expiry (certificate);
+
+  debug (ENSURE, "Certificate %s expires %ld", filename, (long) expires);
+
+  certificate_unref (certificate);
+
+  return expires > time (NULL) + EXPIRY_THRESHOLD;
+}
+
+static bool
+have_certificate (void)
+{
+  char *filename = cockpit_certificate_locate (NULL);
+
+  if (filename == NULL)
+    {
+      debug (ENSURE, "Couldn't locate any certificate");
+      return false;
+    }
+
+  if (strstr (filename, "/0-self-signed.cert"))
+    {
+      debug (ENSURE, "Certificate is self-signed, checking expiry");
+      return check_expiry (filename);
+    }
+
+  debug (ENSURE, "Certificate looks good");
```

Comment:
Completely optional: This could print `filename`.

Reply:
Done.  Thanks for the suggestion.